In [1]:
import pandas as pd
import numpy as np
import sidetable
import sys
import re

import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
sys.path.append('../')
import src.limpieza as lmp

In [3]:
import requests

In [17]:
pd.options.display.max_columns = None

In [24]:
url = 'https://www.ine.es/jaxiT3/files/t/es/csv_bdsc/2078.csv'
res1 = requests.get(url)


In [25]:
res1.status_code

200

In [26]:
from pathlib import Path

data_folder = Path(r"C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data")
file_to_open = data_folder / "2078.csv"
with open(file_to_open, 'wb') as f:
    f.write(res1.content)

In [27]:
df_guiris = pd.read_csv(r"C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data\2078.csv", sep=';', encoding='latin-1', index_col = 0)

In [28]:
df_guiris.shape

(86920, 4)

In [1217]:
df_guiris.head(5)

Viajeros y pernoctaciones             Residencia  \
Puntos turísticos                                                    
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   

                       Periodo   Total  
Puntos turísticos                   
01059 Vitoria-Gasteiz  2022M01  11.759  
01059 Vitoria-Gasteiz  2021M12  16.496  
01059 Vitoria-Gasteiz  2021M11  16.667  
01059 Vitoria-Gasteiz  2021M10  21.467  
01059 Vitoria-Gasteiz  2021M09  18.647

In [1218]:
df_guiris.duplicated().sum()

12322

In [1219]:
df_guiris[df_guiris.duplicated() == True]

Viajeros y pernoctaciones                   Residencia  \
Puntos turísticos                                                          
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero  Residentes en el Extranjero   
02003 Albacete                          Viajero  Residentes en el Extranjero   
...                                         ...                          ...   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   

                       Periodo Total  
Puntos turísticos                 
02003 Albacete         2020M06     .  
02003 Albacete         2020M05     .  
02003 Albacete         2020M04     0  
02003 Albacete         2020M06     .  
02003 Albacete         2020M05     .  
...                        ...   ...  
50297 Zaragoza         2020M05     .  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M06     .  
50297 Zaragoza         2020M05     .  
50297 Zaragoza         2020M04     0  

[12322 rows x 4 columns]

In [1220]:
df_guiris[df_guiris['Total'] == '0']

Viajeros y pernoctaciones                   Residencia  \
Puntos turísticos                                                          
01059 Vitoria-Gasteiz                   Viajero        Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en el Extranjero   
01059 Vitoria-Gasteiz            Pernoctaciones        Residentes en España   
01059 Vitoria-Gasteiz            Pernoctaciones  Residentes en el Extranjero   
02003 Albacete                          Viajero        Residentes en España   
...                                         ...                          ...   
49275 Zamora                     Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                          Viajero        Residentes en España   
50297 Zaragoza                          Viajero  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   

                       Periodo Total  
Puntos turísticos                 
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
02003 Albacete         2020M04     0  
...                        ...   ...  
49275 Zamora           2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  

[425 rows x 4 columns]

Aunque parecía que había duplicados, en la tabla anterior podemos ver que no es así y si resetamos el index desaparecen

In [1221]:
df_guiris.reset_index(inplace=True)

In [1222]:
df_guiris.duplicated().sum()

0

In [1223]:
df_guiris.dtypes

Puntos turísticos        object
Viajeros y pernoctaciones    object
Residencia                   object
Periodo                      object
Total                        object
dtype: object

Tendremos que limpiar algunos nombres

In [1224]:
df_guiris.columns

Index(['Puntos turísticos', 'Viajeros y pernoctaciones', 'Residencia',
       'Periodo', 'Total'],
      dtype='object')

In [1225]:
print(df_guiris['Viajeros y pernoctaciones'].unique())
print(df_guiris['Residencia'].unique())

['Viajero' 'Pernoctaciones']
['Residentes en España' 'Residentes en el Extranjero']


In [1226]:
df_guiris.isnull().sum()

Puntos turísticos           0
Viajeros y pernoctaciones       0
Residencia                      0
Periodo                         0
Total                        5358
dtype: int64

In [1227]:
df_guiris.shape

(86920, 5)

Renombramos las columnas

In [1228]:
df_guiris = df_guiris.rename(columns={'Puntos turísticos': 'Puntos_turisticos'})
columnas_limpias = {col : col.lower().replace(" ", "_") for col in df_guiris.columns}
df_guiris.rename(columns = columnas_limpias, inplace = True)

In [1229]:
df_guiris.columns

Index(['puntos_turisticos', 'viajeros_y_pernoctaciones', 'residencia',
       'periodo', 'total'],
      dtype='object')

Ahora vamos a dejar limpita la columna 'total'. Primero retiraremos los puntos y luego convertiremos los valores en enteros.

Es importante tener en cuenta que en este csv si aparece un punto '.' a secas, significa bien que el dato está protegido por la ley de secreto estadístico, (Artículo 12 de la Ley 7/2006) o en algunas ocasiones que el dato no se recogió. y Para nosotros es como si fuera un nulo.

Esta operación de limpieza tiene por tanto una doble función:
- poder convertir los string en int
- convertir los '.' en NaN para manejarlos después

In [1230]:
df_guiris['total'] = df_guiris['total'].str.replace('.', '')

In [1231]:
df_guiris['total'] = pd.to_numeric(df_guiris['total'], errors='coerce').astype('Int64')

Podemos comprobar que los nulos han aumentado de 5358 a 13056

In [1232]:
df_guiris.isnull().sum()

puntos_turisticos                0
viajeros_y_pernoctaciones        0
residencia                       0
periodo                          0
total                        13056
dtype: int64

Sabemos por la documentación del csv que entre abril y junio del 2020 no hubo recogida de datos a causa de la crisis del covid-19. Veamos cómo están esos datos.

In [1233]:
df_guiris[(df_guiris['periodo'] == '2020M04') | (df_guiris['periodo'] == '2020M05') | (df_guiris['periodo'] == '2020M06')]

puntos_turisticos viajeros_y_pernoctaciones  \
19     01059 Vitoria-Gasteiz                   Viajero   
20     01059 Vitoria-Gasteiz                   Viajero   
21     01059 Vitoria-Gasteiz                   Viajero   
224    01059 Vitoria-Gasteiz                   Viajero   
225    01059 Vitoria-Gasteiz                   Viajero   
...                      ...                       ...   
86530         50297 Zaragoza            Pernoctaciones   
86531         50297 Zaragoza            Pernoctaciones   
86734         50297 Zaragoza            Pernoctaciones   
86735         50297 Zaragoza            Pernoctaciones   
86736         50297 Zaragoza            Pernoctaciones   

                        residencia  periodo  total  
19           Residentes en España  2020M06   <NA>  
20           Residentes en España  2020M05   <NA>  
21           Residentes en España  2020M04      0  
224    Residentes en el Extranjero  2020M06   <NA>  
225    Residentes en el Extranjero  2020M05   <NA>  
...                            ...      ...    ...  
86530        Residentes en España  2020M05   <NA>  
86531        Residentes en España  2020M04      0  
86734  Residentes en el Extranjero  2020M06   <NA>  
86735  Residentes en el Extranjero  2020M05   <NA>  
86736  Residentes en el Extranjero  2020M04      0  

[1272 rows x 5 columns]

Dado que en los meses del confinamiento toda la actividad turística estuvo suspendida, vamos a sustituir los nulos por 0, pues no es tanto que no hubiese recogida de datos como que no había turistas.

In [1234]:
df_guiris.loc[(df_guiris['periodo'] == '2020M04') | 
              (df_guiris['periodo'] == '2020M05') | 
              (df_guiris['periodo'] == '2020M06'), 'total'] = 0

Y comprobamos que los nulos han descendido de 13056 a 12208

In [1235]:
df_guiris.isnull().sum()

puntos_turisticos                0
viajeros_y_pernoctaciones        0
residencia                       0
periodo                          0
total                        12208
dtype: int64

Los nulos que quedan se deben a que o bien no había recogida de datos en esa fecha para esa zona turística, o bien el dato no está publicado porque está protegido por secreto estadístico.
En todo caso no se pueden eliminar alegremente. En el caso de que falten en meses puntuales pero dispongamos de datos para el resto del año, quizá convendría sustiruirlos.

In [1236]:
null_rows = pd.isnull(df_guiris['total'])

In [1237]:
df_guiris[null_rows].to_csv('nulls.csv')

In [1238]:
df_guiris['periodo'] = df_guiris['periodo'].str.replace('M', '-')

In [1239]:
df_guiris.to_pickle('guiris.pkl')
#df_guiris = pd.read_pickle('guiris.pkl')

In [1240]:
df_guiris['periodo'] = pd.to_datetime(df_guiris['periodo'], format='%Y%m%')
#df_guiris['periodo'] = df_guiris['periodo'].dt.date.apply(lambda x: x.strftime('%Y-%m'))

In [1241]:
df_guiris.dtypes

puntos_turisticos                    object
viajeros_y_pernoctaciones            object
residencia                           object
periodo                      datetime64[ns]
total                                 Int64
dtype: object

In [1242]:
df_guiris.set_index('periodo', inplace=True)

Renombramos los valores

In [1243]:
df_guiris['residencia'].replace({'Residentes en España': "residentes_espana", 'Residentes en el Extranjero': 'residentes_en_el_extranjero'}, inplace=True)
df_guiris['viajeros_y_pernoctaciones'] = df_guiris['viajeros_y_pernoctaciones'].str.lower()

In [1244]:
lista = ['Barcelona']

In [1245]:
lmp.pastor_de_guiris(df_guiris, lista)

TypeError: unhashable type: 'list'